core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1168,6 +1168,61 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    */
   final def tracingStatus(flag: TracingStatus): ZIO[R, E, A] = new ZIO.TracingStatus(self, flag)
 
+  /**
+   * Applies `f` if the predicate fails.
+   */
+  final def filterOrElse[R1 <: R, E1 >: E, A1 >: A](p: A => Boolean)(f: A => ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
+    self.flatMap {
+      case v if !p(v) => f(v)
+      case v          => ZIO.succeed(v)
+    }
+
+  /**
+   * Supplies `zio` if the predicate fails.
+   */
+  final def filterOrElse_[R1 <: R, E1 >: E, A1 >: A](p: A => Boolean)(zio: => ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
+    filterOrElse[R1, E1, A1](p)(_ => zio)
+
+  /**
+   * Fails with `e` if the predicate fails.
+   */
+  final def filterOrFail[E1 >: E](p: A => Boolean)(e: => E1): ZIO[R, E1, A] =
```

Comment:
The only issue I see is with the parameter ordering for `filterOrElse_` and `filterOrElse`. Ordinarily, the error would be first, because it is most likely to be constant, while the predicate is most likely to vary. 

Is there a reason you have the current ordering?

Reply:
Entirely matching the parameter ordering of the Scala stdlib as well as cats. I'm opening to swapping them if you'd prefer.

Reply:
Works for me!